workflow/algorithm_test.go

```diff
@@ -0,0 +1,268 @@
+			{Key: "nodeKey1"},
+			{Key: "nodeKey2"},
+			{Key: "nodeKey3"},
+			{Key: "nodeKey4"},
+			{Key: "nodeKey5"},
+			{Key: "nodeKey6"},
+			{Key: "nodeKey7"}},
+		Edges: []Edge{
+			{Src: "nodeKey1", Dst: "nodeKey2"},
+			{Src: "nodeKey2", Dst: "nodeKey3"},
+			{Src: "nodeKey2", Dst: "nodeKey4"},
+			{Src: "nodeKey3", Dst: "nodeKey5"},
+			{Src: "nodeKey4", Dst: "nodeKey6"},
+			{Src: "nodeKey4", Dst: "nodeKey7"},
+		},
+	}
+}
+
+func TestChildrenIDs(t *testing.T) {
+	var tests = []struct {
+		graph    *Workflow
+		node     string
+		children []string
+	}{
+		{graph: defaultGraph(), node: "nodeKey1", children: []string{"nodeKey2"}},
+		{graph: defaultGraph(), node: "nodeKey2", children: []string{"nodeKey3", "nodeKey4"}},
+		{graph: defaultGraph(), node: "nodeKey3", children: []string{"nodeKey5"}},
+		{graph: defaultGraph(), node: "nodeKey4", children: []string{"nodeKey6", "nodeKey7"}},
+		{graph: defaultGraph(), node: "nodeKey5", children: []string{}},
+		{graph: defaultGraph(), node: "nodeKey6", children: []string{}},
+		{graph: defaultGraph(), node: "nodeKey7", children: []string{}},
+	}
+	for _, test := range tests {
+		assert.Equal(t, test.graph.ChildrenIDs(test.node), test.children)
+	}
+}
+
+func TestParentIDs(t *testing.T) {
+	var tests = []struct {
+		graph   *Workflow
+		node    string
+		parents []string
+	}{
+		{graph: defaultGraph(), node: "nodeKey1", parents: []string{}},
+		{graph: defaultGraph(), node: "nodeKey2", parents: []string{"nodeKey1"}},
+		{graph: defaultGraph(), node: "nodeKey3", parents: []string{"nodeKey2"}},
+		{graph: defaultGraph(), node: "nodeKey4", parents: []string{"nodeKey2"}},
+		{graph: defaultGraph(), node: "nodeKey5", parents: []string{"nodeKey3"}},
+		{graph: defaultGraph(), node: "nodeKey6", parents: []string{"nodeKey4"}},
+		{graph: defaultGraph(), node: "nodeKey7", parents: []string{"nodeKey4"}},
+	}
+	for _, test := range tests {
+		assert.Equal(t, test.graph.ParentIDs(test.node), test.parents)
+	}
+}
+
+func TestFindNode(t *testing.T) {
+	var tests = []struct {
+		graph   *Workflow
+		node    string
+		present bool
+	}{
+		{graph: defaultGraph(), node: "nodeKey1", present: true},
+		{graph: defaultGraph(), node: "nodeKey2", present: true},
+		{graph: defaultGraph(), node: "nodeKey3", present: true},
+		{graph: defaultGraph(), node: "nodeKey4", present: true},
+		{graph: defaultGraph(), node: "nodeKey5", present: true},
+		{graph: defaultGraph(), node: "nodeKey6", present: true},
+		{graph: defaultGraph(), node: "nodeKey7", present: true},
+		{graph: defaultGraph(), node: "nodeKey8", present: false},
+	}
+	for _, test := range tests {
+		node, err := test.graph.FindNode(test.node)
+		if test.present {
+			assert.NoError(t, err)
+			assert.Equal(t, node.Key, test.node)
+		} else {
+			assert.Error(t, err)
+		}
+	}
+}
+
+func TestHasNodes(t *testing.T) {
+	var tests = []struct {
+		graph *Workflow
+		null  bool
+	}{
+		{graph: defaultGraph(), null: false},
+		{graph: &Workflow{}, null: true},
```

Comment:
again about hasNodes...
```suggestion
		{graph: &Workflow{}, hasNodes: false},
```